### ISSUE TITLE
Traceability Matrix

### ISSUE THREAD

**Issue by smiths** [smiths/swhs#4]
Please check the Traceability Matrix in Section 7 and correct as necessary.  For each row a tick goes in each column where the entity in the row depends on the corresponding column.  For instance GD2 depends on T1, A3, A4, A5 and A6.  (The current matrix also shows a dependence on GD1, but I don't think this is correct.)  A change in any of these dependencies likely means that GD2 would also have to change.
Ideally the matrix should have cross reference links to the main document for easy navigation.  Also, it would be great if all of the traceable entries could be in the matrix, including requirements (R?) and Likely Changes (LC?).  We can put the matrix in landscape mode to create more room.  It also could be split into separate traceability matricies.  For example, the likely changes only trace to assumptions, so this could be in a separate table.

**Comment by bmaclach** [comment]
Committed some changes addressing these issues. I split the single matrix into three matrices so that everything could be included, and corrected any errors. However, I'm not confident in what I've done with regards to traceability of the requirements. I wasn't quite sure how to interpret that. 

Specifically, which items should we be looking at in relation to the requirements? I settled on instance models, since they are the primary item referenced in the requirements themselves, but then I was left unsure about the perspective from which the traceability should be viewed. By that I mean, do we say that the requirements depend on the instance models, or that the instance models depend on the requirements? Neither way seems right to me. It doesn't seem right to have the software requirements depend on the models, because the models are developed in order to meet requirements. But then again, it would make sense to say that the _fulfillment_ of the requirements depends upon the models. Is that the way I should be looking at it?

Currently, I've structured the matrix to show the models depending on the requirements, but as I mentioned before this doesn't make much sense to me either. For example, requirement 4 involves values from IM1 and IM2. But if requirement 4 were to change, IM1 and IM2 would still be completely correct. Similar comments could be made about the rest of the requirements, too. Which once again makes me think that I should be viewing this as "which items do the requirements depend upon in order to be fulfilled?"

In other words, I may have sorted this out in my head as I typed this. But I'd still like to hear what you think before I change anything.

**Comment by smiths** [comment]
Your thoughts in the second to last paragraph are correct.  An X in a column means that the item in the corresponding row depends on the entry in the column.  Therefore, I think you want something close to the transpose of your Table 5.  Your rows should be the requirements and the columns are the instance models.  This way you can say that R2 depends on R1.  (R2 doesn't actually depend on IM1, IM2, IM3 and IM4.  Actually, IM1 to IM4 depend on R2.)  If R1 is not implemented correctly, than R2 cannot be correct.  If R1 changes, then R2 might have to change.  Since R4 depends on IM1, IM2, R1 and R2, you should also have columns for R1 and R2.
If you change Table 5 to the "depends on" view, it will be consistent with your other tables.
Your tables would be used to determine the consequences of change.  If A3 changes, we see that GD2 and DD2 might need to be updated.  If GD2 is changed, then IM1 and IM2 might have to be modified.  The tables provide a means to visualize the consequences of change.
R3 does not look like it depends on anything, which is always a cause for consideration.  Thinking about it, R3 depends on the data constraints, which are not currently represented in the traceability tables.  What do you think about adding a column for data constraints, to make this traceability explicit?
If you have time, do you think you could draw a "depends on" graph?  That might convey the traceability information better than the traceability matrix.  If you are looking for a drawing tool, LucidChart (a free version is available) is well suited for drawing graphs.

**Comment by bmaclach** [comment]
Thanks for the response, what you've described makes sense. The one confusing part is where you say that IM1 to IM4 depend on R2. If that is the case, should the instance models be included as rows as well as columns? That way we can show that the instance models depend on R2 and also show requirement dependencies on the models, such as how R4 depends on IM1 and IM2.

I should have time to make a graph.

**Comment by smiths** [comment]
Yes, the IMs can be rows as well as columns.  What you describe sounds good to me.  
I'm glad to hear that you will tackle drawing a graph.  If you can create a graph, this might give us clearer picture of how things fit together.

**Comment by bmaclach** [comment]
I made some charts and added them to the SRS. I split the chart up into three parts, similar to how we had three matrices before. I am not sure if that was what you had in mind, but including everything in one chart would involve so many arrows that I think it would be very hard to follow. Even with it split into three charts, I am worried that the arrows can be harder to follow than a column in a table.

If you want me to switch back to the matrices, or if you had something else in mind for the charts, let me know.

**Comment by smiths** [comment]
The graphs look great.  (Can you please modify the text to call them graphs, not charts.  I mean graph in the computer science sense (graph theory)).  
I would like to have both the traceability matrix and the graphs.  I think both views of the knowledge have something to offer.  Can you please put the traceability matrices back in the SRS?
Can you please combine Figure 2 and 3?  I know it will be messy, but as it is, Figure 2 only tells part of the story.  I feel like a pattern of sorts might arise by looking at the full graph.  Figure 4 is mostly disjoint from the other graph, so you can leave it independent.
In the introductory blurb, can you please say the meaning of the arrow.  I don't think it is a "depends on" relation, but rather a "is depended on by" relation.  For instance, A2 is depended on by GD1.  As you say, if a component changes, any of the components it points to might also have to be changed.
Good work!

**Comment by bmaclach** [comment]
Thanks for the feedback. I can make those changes. I wasn't familiar with the computer science sense of graphs... good to know!

**Comment by bmaclach** [comment]
Added back the matrices, combined Figures 2 and 3, and adjusted the explanation for the graphs to reflect the meaning of the arrows.

**Comment by smiths** [comment]
Looks great.  Thank you Brooks.
